Only in StRoot/StRichPIDMaker: .consign
diff -urb StRoot_old/StRichPIDMaker/StRichPIDMaker.cxx StRoot/StRichPIDMaker/StRichPIDMaker.cxx
--- StRoot_old/StRichPIDMaker/StRichPIDMaker.cxx	Fri Feb 22 09:34:25 2002
+++ StRoot/StRichPIDMaker/StRichPIDMaker.cxx	Tue Nov 19 13:35:52 2002
@@ -340,7 +340,8 @@
 
   //
   // default version
-  mProductionVersion = -999;
+  // No, this is the global thing
+  mProductionVersion = 20013;
 
   //
   // initialize cuts, values and parameters for processing
@@ -365,7 +366,10 @@
     mPtCut        = 0.5*GeV; // GeV/c
     mEtaCut       = 0.5; 
     mLastHitCut   = 100.0*centimeter;
-    mDcaCut       = 3.0*centimeter;
+//    mDcaCut       = 3.0*centimeter;
+// Remove DCA cut
+    mDcaCut = 10000.*centimeter;
+    
     mFitPointsCut = 20;
     mPathCut      = 500*centimeter;
     mPadPlaneCut  = 1.0*centimeter;
@@ -996,9 +1000,19 @@
     
     mNumberOfPrimaries = tempEvent->primaryVertex()->numberOfDaughters();  
     
-    for (int ii=0; ii<mNumberOfPrimaries; ii++)  {
-
-	StTrack* track = tempEvent->primaryVertex()->daughter(ii);
+//      for (int ii=0; ii<mNumberOfPrimaries; ii++)  {
+//	StTrack* track = tempEvent->primaryVertex()->daughter(ii);
+    StSPtrVecTrackNode& theNodes = tempEvent->trackNodes();
+    for (StSPtrVecTrackNodeIterator titer = theNodes.begin();
+	 titer != theNodes.end(); ++titer) {
+	for (unsigned int ientry = 0; ientry < (*titer)->entries();
+	     ++ientry) {
+	    
+	    StTrack* track = (*titer)->track(ientry);
+	    if (!track) {
+		cout << "Huh?  No track " << endl;
+		continue;
+	    }
 
 	if (this->checkTrack(track)) {
 	    
@@ -1090,11 +1104,13 @@
 	    }
 	}
 	//	if (tempTrack->geometry()->charge() < 0) {
-	if (tempEvent->primaryVertex()->daughter(ii)->geometry()->charge() < 0) {
-	    mNegativePrimaries++;
+//  	if (tempEvent->primaryVertex()->daughter(ii)->geometry()->charge() < 0) {
+//  	    mNegativePrimaries++;
+//  	}
 	}
     }
  
+ 
     //
     // make monte carlo associations 
     //
Only in StRoot/StRichPIDMaker: StRichPIDMaker.cxx.flc
